subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionCache.kt

```diff
@@ -0,0 +1,142 @@
+import org.gradle.internal.Factory
+import org.gradle.internal.concurrent.Stoppable
+import org.gradle.internal.file.FileAccessTimeJournal
+import org.gradle.internal.resource.local.SingleDepthFileAccessTracker
+import java.io.File
+
+
+internal
+class InstantExecutionCache(
+    private val cacheRepository: CacheRepository,
+    private val cacheCleanupFactory: CleanupActionFactory,
+    private val fileAccessTimeJournal: FileAccessTimeJournal,
+    private val startParameter: InstantExecutionStartParameter
+) : Stoppable {
+
+    private
+    val cacheRootDir
+        get() = startParameter.rootDirectory.resolve(".gradle/configuration-cache")
+
+    private
+    val cleanupDepth = 1
+
+    private
+    val cleanupMaxAgeDays = LeastRecentlyUsedCacheCleanup.DEFAULT_MAX_AGE_IN_DAYS_FOR_RECREATABLE_CACHE_ENTRIES
+
+    private
+    val cache by unsafeLazy {
+        cacheRepository.configurationCache()
+            .withOnDemandLockMode() // Don't need to lock anything until we use the caches
+            .withLruCacheCleanup()
+            .open()
+    }
+
+    private
+    fun CacheRepository.configurationCache() =
+        cache(cacheRootDir).withDisplayName("Configuration Cache")
+
+    private
+    fun CacheBuilder.withOnDemandLockMode() =
+        withLockOptions(LockOptionsBuilder.mode(FileLockManager.LockMode.OnDemand))
+
+    private
+    fun CacheBuilder.withLruCacheCleanup(): CacheBuilder =
+        withCleanup(cacheCleanupFactory.create(LeastRecentlyUsedCacheCleanup(
+            SingleDepthFilesFinder(cleanupDepth),
+            fileAccessTimeJournal,
+            cleanupMaxAgeDays
+        )))
+
+    override fun stop() {
+        cache.close()
+    }
+
+    private
+    val fileAccessTracker by unsafeLazy {
+        SingleDepthFileAccessTracker(fileAccessTimeJournal, cacheRootDir, cleanupDepth)
+    }
+
+    sealed class CheckedFingerprint {
+        object NotFound : CheckedFingerprint()
+        object Valid : CheckedFingerprint()
+        class Invalid(val reason: String) : CheckedFingerprint()
+    }
+
+    fun useForFingerprintCheck(cacheKey: String, check: (File) -> String?): CheckedFingerprint =
+        cache.withFileLock(Factory {
+            val fingerprint = cache.baseDirFor(cacheKey).fingerprintFile
+            when {
+                !fingerprint.isFile -> CheckedFingerprint.NotFound
+                else -> {
+                    when (val invalidReason = check(fingerprint)) {
+                        null -> {
+                            fileAccessTracker.markAccessed(fingerprint) // Prevent cleanup before state load
+                            CheckedFingerprint.Valid
+                        }
+                        else -> CheckedFingerprint.Invalid(invalidReason)
+                    }
+                }
+            }
+        })
+
+    fun useForStateLoad(cacheKey: String, action: (File) -> Unit) {
+        cache.withFileLock {
```

Comment:
⭕ The usages of `cache.withFileLock { ... cache.baseDirFor(cacheKey) ... }`  could be deduped with a `withBaseCacheDirFor(cacheKey) { ... }`

Reply:
Done